client-sdk/ts-web/core/src/hdkey.ts

```diff
@@ -0,0 +1,112 @@
+import {sha512} from 'js-sha512';
+import {SignKeyPair, sign} from 'tweetnacl';
+import {generateMnemonic, mnemonicToSeed} from 'bip39';
+import {concat} from './misc';
+
+const ED25519_CURVE = 'ed25519 seed';
+const HARDENED_OFFSET = 0x80000000;
+const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
+
+/**
+ * HDKey handles hierarchical key generation according to ADR 0008
+ * https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0008-standard-account-key-generation.md
+ */
+export class HDKey {
+    public readonly keypair: SignKeyPair;
+
+    constructor(public readonly privateKey: Uint8Array, public readonly chainCode: Uint8Array) {
+        this.keypair = sign.keyPair.fromSeed(privateKey);
+    }
+
+    /**
+     * Generates the keypair matching the supplied parameters
+     * @param mnemonic BIP-0039 Mnemonic
+     * @param index Account index
+     * @param passphrase Optional BIP-0039 passphrase
+     * @returns SignKeyPair for these parameters
+     */
+    public static async getAccountSigner(
+        mnemonic: string,
+        index: number = 0,
+        passphrase?: string,
+    ): Promise<SignKeyPair> {
+        const key = await HDKey.fromMnemonic(mnemonic, passphrase);
+        return key.derivePath(`m/44'/474'/${index}'`).keypair;
+    }
+
+    /**
+     * Returns an instance of HDKey that can then be used for derivation
+     * @param seed Master seed for key derivation
+     * @returns Instance of HDKey
+     */
+    public static fromSeed(seed: Uint8Array): HDKey {
```

Comment:
Can we make this function match the Go implementation and reject seeds smaller than 64 bytes?

Reference: https://github.com/oasisprotocol/oasis-core/blob/27c2ceaccfa696351c7d7ca89645e3887fac8d4f/go/common/crypto/slip10/slip10.go#L28-L43

Note that this is actually what [SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md#master-key-generation) specifies, but their [_Test vector 1 for ed25519_](https://github.com/satoshilabs/slips/blob/master/slip-0010.md#test-vector-1-for-ed25519) violates that :slightly_smiling_face: .

Reply:
By specification, SLIP-0010 is supposed to be compatible with BIP-0032 :

> Generate a seed byte sequence S of a chosen length (between 128 and 512 bits; 256 bits is advised) from a (P)RNG.

These vectors, being 128 bits, are to test that the underlying implementation works with older 128 bits BIP-0032 keys. 

If the key derivation happens through BIP-0039, the seed will be 512 bits anyways, but I don't think that we should exclude users that want to reuse existing seeds if they absolutely want to

Reply:
> These vectors, being 128 bits, are to test that the underlying implementation works with older 128 bits BIP-0032 keys.

There is nothing in the SLIP-0010 spec that allows for this, apart from `It uses a process similar and compatible to BIP-0032. For other curves it uses a different salt than BIP-0032.` and `Curve = "Bitcoin seed" for the secp256k1 curve (this is compatible to BIP-0032).`.

On the other hand:

SLIP-0010: `Generate a seed byte sequence S of 512 bits according to BIP-0039.`

BIP-0032: `Generate a seed byte sequence S of a chosen length (between 128 and 512 bits; 256 bits is advised) from a (P)RNG.`

> After thinking about it some more, I strongly think that deviating from all the other implementations and libraries out there, and cutting half of the test vectors would be problematic. It is ultimately up to the user to use the sdk as they see fit, and here it would be similar to refusing 3 words or 6 words mnemonics : while it's not optimal/recommended, it's still valid and the sdk should not be so opinionated that it refuses a valid key from a user

I disagree.

 * It is possible to still test the other test vector.  If typescript doesn't have a way to test package private functions, it might be more annoying to do so, but I do not write typescript ever.
 * ADR-0008 mandates BIP-0039, which produces 512 bit `S`.
 * SLIP-0010 explicitly defines 512 bit `S` (BIP-0032 explicitly does not, but we are not using BIP-0032, we are using SLIP-0010).
 * SLIP-0010's definition of the algorithm runs contrary to their compatibility claims, and the algorithm should be implemented as specified.

In general, I tend to favor the most concrete/narrow reading/interpretation of a specification possible because not doing so leads to problems down the lines.  Because it would helps clarify things, amending ADR-0008 to point out that SLIP-0010 is poorly specified and self contradictory, along with the interpretation we use is fine.

Reply:
cosmos1hkmy9n3mg5j8t4fsvvm7zzw4x0lar5n3eu2tt9